File: alphazero/core/mcts/transposition_table.cpp
#include <random>

#include "transposition_table.h"
#include "mcts_node.h"

namespace alphazero {

TranspositionTable::TranspositionTable(size_t max_size)
    : max_size_(max_size) {
}

TranspositionTable::~TranspositionTable() {
    // Note: We don't delete the nodes since they are owned by the MCTS tree
}

MCTSNode* TranspositionTable::lookup(uint64_t hash_value) {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    
    auto it = table_.find(hash_value);
    if (it != table_.end()) {
        return it->second;
    }
    
    return nullptr;
}

void TranspositionTable::store(uint64_t hash_value, MCTSNode* node) {
    if (node == nullptr) {
        return; // Don't store null nodes
    }
    
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    
    // If table is full, evict an existing entry
    if (table_.size() >= max_size_) {
        // Generate a random index (thread-safe)
        static thread_local std::mt19937 gen(std::random_device{}());
        std::uniform_int_distribution<size_t> dist(0, table_.size() - 1);
        
        // Find a random entry to evict
        auto it = table_.begin();
        std::advance(it, dist(gen) % table_.size()); // Ensure we don't go out of bounds
        table_.erase(it);
    }
    
    // Store the new node
    table_[hash_value] = node;
}

void TranspositionTable::clear() {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    table_.clear();
}

size_t TranspositionTable::size() const {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    return table_.size();
}

bool TranspositionTable::contains(uint64_t hash_value) const {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    return table_.find(hash_value) != table_.end();
}

// LRU Transposition Table implementation

LRUTranspositionTable::LRUTranspositionTable(size_t max_size)
    : TranspositionTable(max_size) {
}

MCTSNode* LRUTranspositionTable::lookup(uint64_t hash_value) {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    
    auto it = table_.find(hash_value);
    if (it != table_.end()) {
        // Make sure the node still exists and hasn't been deallocated
        if (it->second != nullptr) {
            // Check if hash exists in lru_map before trying to access it
            auto lru_map_it = lru_map_.find(hash_value);
            if (lru_map_it != lru_map_.end()) {
                // Move the hash to the front of the LRU list
                auto lru_it = lru_map_it->second;
                lru_list_.erase(lru_it);
                lru_list_.push_front(hash_value);
                lru_map_[hash_value] = lru_list_.begin();
            } else {
                // If the hash is in table_ but not in lru_map_, add it
                lru_list_.push_front(hash_value);
                lru_map_[hash_value] = lru_list_.begin();
            }
            
            return it->second;
        } else {
            // If the node is null, remove the entry from the table
            auto lru_map_it = lru_map_.find(hash_value);
            if (lru_map_it != lru_map_.end()) {
                lru_list_.erase(lru_map_it->second);
                lru_map_.erase(lru_map_it);
            }
            table_.erase(it);
            return nullptr;
        }
    }
    
    return nullptr;
}

void LRUTranspositionTable::store(uint64_t hash_value, MCTSNode* node) {
    if (node == nullptr) {
        return; // Don't store null nodes
    }
    
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    
    // If hash already exists, update it
    auto it = table_.find(hash_value);
    if (it != table_.end()) {
        // Only update if the new node has more information
        if (node->visit_count.load() > it->second->visit_count.load()) {
            it->second = node;
        }
        
        // Update LRU position regardless
        auto lru_map_it = lru_map_.find(hash_value);
        if (lru_map_it != lru_map_.end()) {
            lru_list_.erase(lru_map_it->second);
            lru_list_.push_front(hash_value);
            lru_map_[hash_value] = lru_list_.begin();
        } else {
            // If somehow the hash is in table_ but not in lru_map_, add it
            lru_list_.push_front(hash_value);
            lru_map_[hash_value] = lru_list_.begin();
        }
        
        return;
    }
    
    // If table is full, remove the least recently used entry
    if (table_.size() >= max_size_ && !lru_list_.empty()) {
        uint64_t lru_hash = lru_list_.back();
        table_.erase(lru_hash);
        lru_map_.erase(lru_hash);
        lru_list_.pop_back();
    }
    
    // Add the new entry
    table_[hash_value] = node;
    lru_list_.push_front(hash_value);
    lru_map_[hash_value] = lru_list_.begin();
}

void LRUTranspositionTable::clear() {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    table_.clear();
    lru_list_.clear();
    lru_map_.clear();
}

} // namespace alphazero

File: alphazero/python/mcts/transposition_table.py
import threading
from typing import Dict, Any, Optional, Tuple

class TranspositionTable:
    """
    Transposition table for MCTS to reuse evaluations of previously seen states.
    
    This is a simple hash table that maps game state hashes to nodes in the search tree.
    Thread-safe implementation for parallel MCTS.
    """
    def __init__(self, max_size: int = 1000000):
        """
        Initialize the transposition table.
        
        Args:
            max_size: Maximum number of entries in the table
        """
        self.max_size = max_size
        self.table: Dict[int, Any] = {}
        self.lock = threading.RLock()
    
    def lookup(self, hash_value: int) -> Optional[Any]:
        """
        Look up a node in the table.
        
        Args:
            hash_value: Hash value of the game state
            
        Returns:
            The node if found, None otherwise
        """
        with self.lock:
            return self.table.get(hash_value)
    
    def store(self, hash_value: int, node: Any) -> None:
        """
        Store a node in the table.
        
        Args:
            hash_value: Hash value of the game state
            node: Node to store
        """
        with self.lock:
            # If table is full, don't add new entries
            if len(self.table) >= self.max_size:
                return
            
            self.table[hash_value] = node
    
    def clear(self) -> None:
        """Clear the table."""
        with self.lock:
            self.table.clear()
    
    def size(self) -> int:
        """Return the number of entries in the table."""
        with self.lock:
            return len(self.table)
    
    def __contains__(self, hash_value: int) -> bool:
        """Check if a hash value is in the table."""
        with self.lock:
            return hash_value in self.table


class LRUTranspositionTable(TranspositionTable):
    """
    Transposition table with LRU (Least Recently Used) eviction policy.
    
    This is an enhanced version of the transposition table that removes
    the least recently used entries when the table is full.
    """
    def __init__(self, max_size: int = 1000000):
        """
        Initialize the LRU transposition table.
        
        Args:
            max_size: Maximum number of entries in the table
        """
        super().__init__(max_size)
        self.access_count = 0
        self.access_time: Dict[int, int] = {}
    
    def lookup(self, hash_value: int) -> Optional[Any]:
        """
        Look up a node in the table and update its access time.
        
        Args:
            hash_value: Hash value of the game state
            
        Returns:
            The node if found, None otherwise
        """
        with self.lock:
            node = self.table.get(hash_value)
            if node is not None:
                self.access_count += 1
                self.access_time[hash_value] = self.access_count
            return node
    
    def store(self, hash_value: int, node: Any) -> None:
        """
        Store a node in the table, evicting the least recently used entry if necessary.
        
        Args:
            hash_value: Hash value of the game state
            node: Node to store
        """
        with self.lock:
            # If table is full, remove the least recently used entry
            if len(self.table) >= self.max_size:
                # Find the least recently used entry
                lru_hash = min(self.access_time.items(), key=lambda x: x[1])[0]
                del self.table[lru_hash]
                del self.access_time[lru_hash]
            
            # Store the new entry
            self.table[hash_value] = node
            self.access_count += 1
            self.access_time[hash_value] = self.access_count
    
    def clear(self) -> None:
        """Clear the table and access times."""
        with self.lock:
            self.table.clear()
            self.access_time.clear()
            self.access_count = 0